src/Time/TimeSteppers/DormandPrince5.hpp

```diff
@@ -0,0 +1,260 @@
+                                            11.0 / 84.0}};
+  static constexpr std::array<double, 5> c{{0.2, 0.3, 0.8, 8.0 / 9.0, 1.0}};
+
+  // Coefficients for dense output, taken from Sec. 7.2 of
+  // \cite NumericalRecipes
+  static constexpr std::array<double, 6> d{
+      {-12715105075.0 / 11282082432.0, 87487479700.0 / 32700410799.0,
+       -10690763975.0 / 1880347072.0, 701980252875.0 / 199316789632.0,
+       -1453857185.0 / 822651844.0, 69997945.0 / 29380423.0}};
+
+  DormandPrince5() = default;
+  DormandPrince5(const DormandPrince5&) noexcept = default;
+  DormandPrince5& operator=(const DormandPrince5&) noexcept = default;
+  DormandPrince5(DormandPrince5&&) noexcept = default;
+  DormandPrince5& operator=(DormandPrince5&&) noexcept = default;
+  ~DormandPrince5() noexcept override = default;
+
+  template <typename Vars, typename DerivVars>
+  void update_u(gsl::not_null<Vars*> u,
+                gsl::not_null<History<Vars, DerivVars>*> history,
+                const TimeDelta& time_step) const noexcept;
+
+  template <typename Vars, typename DerivVars>
+  void dense_update_u(gsl::not_null<Vars*> u,
+                      const History<Vars, DerivVars>& history,
+                      double time) const noexcept;
+
+  uint64_t number_of_substeps() const noexcept override;
+
+  size_t number_of_past_steps() const noexcept override;
+
+  double stable_step() const noexcept override;
+
+  TimeStepId next_time_id(const TimeStepId& current_id,
+                          const TimeDelta& time_step) const noexcept override;
+
+  template <typename Vars, typename DerivVars>
+  bool can_change_step_size(
+      const TimeStepId& time_id,
+      const TimeSteppers::History<Vars, DerivVars>& /*history*/) const
+      noexcept {
+    return time_id.substep() == 0;
+  }
+
+  WRAPPED_PUPable_decl_template(DormandPrince5);  // NOLINT
+
+  explicit DormandPrince5(CkMigrateMessage* /*unused*/) noexcept {}
+
+  // clang-tidy: do not pass by non-const reference
+  void pup(PUP::er& p) noexcept override {  // NOLINT
+    TimeStepper::Inherit::pup(p);
+  }
+};
+
+inline bool constexpr operator==(const DormandPrince5& /*lhs*/,
+                                 const DormandPrince5& /*rhs*/) noexcept {
+  return true;
+}
+
+inline bool constexpr operator!=(const DormandPrince5& /*lhs*/,
+                                 const DormandPrince5& /*rhs*/) noexcept {
+  return false;
+}
+
+template <typename Vars, typename DerivVars>
+void DormandPrince5::update_u(
+    const gsl::not_null<Vars*> u,
+    const gsl::not_null<History<Vars, DerivVars>*> history,
+    const TimeDelta& time_step) const noexcept {
+  const size_t substep = history->size() - 1;
+  const auto& u0 = history->begin().value();
+  const double dt = time_step.value();
+
+  switch (substep) {
+    case 0: {
+      *u += (a2 * dt) * history->begin().derivative();
+      break;
+    }
+    case 1: {
+      *u = u0;
+      for (size_t i = 0; i < 2; ++i) {
```

Comment:
What about adding a lambda at the beginning of the function:
```cpp
[&u0, &history](const auto& coeffs) noexcept {
  for (size_t i = 0; i < coeffs.size(); ++i) {
    *u += (gsl::at(coeffs, i) * dt) * ...;
  }
};
```
and then using an `if/else` approach instead of switches. Seems like that could also same a lot of code here.

Reply:
Good idea.  In fact, that is basically what is done in ABN, except for the different orders instead of the substeps.  (Although it uses a special utility to do the sum and I don't know why I thought that was better than just a for loop.)